File: examples/concurrent-browser-sessions/rest/csharp/ConcurrentScreenshots.cs
// Captures screenshots of multiple pages concurrently using the Browserless REST API.
// Uses System.Net.Http.HttpClient with Task.WhenAll — no extra packages needed.
//
// Run: dotnet run
//
// Note: for production use, replace the string-interpolated JSON body with
// System.Text.Json.JsonSerializer.Serialize to handle URLs containing quotes
// or backslashes safely.

using System;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;

class ConcurrentScreenshots
{
    static async Task Main()
    {
        const string token = "YOUR_API_TOKEN_HERE";
        string[] urls =
        {
            "https://scraping-sandbox.netlify.app/products",
            "https://scraping-sandbox.netlify.app/contact-us",
            "https://scraping-sandbox.netlify.app/receipt",
            "https://scraping-sandbox.netlify.app/dashboard",
            "https://scraping-sandbox.netlify.app/helix",
        };

        using var client = new HttpClient();

        await Task.WhenAll(urls.Select(async (url, i) =>
        {
            var body = new StringContent(
                $"{{\"url\":\"{url}\",\"options\":{{\"type\":\"png\",\"fullPage\":true}}}}",
                Encoding.UTF8,
                "application/json"
            );
            var response = await client.PostAsync(
                $"https://production-sfo.browserless.io/screenshot?token={token}",
                body
            );
            var bytes = await response.Content.ReadAsByteArrayAsync();
            await File.WriteAllBytesAsync($"screenshot-{i + 1}.png", bytes);
            Console.WriteLine($"Saved screenshot-{i + 1}.png");
        }));

        Console.WriteLine("All done");
    }
}

File: examples/concurrent-browser-sessions/rest/java/ConcurrentScreenshots.java
// Captures screenshots of multiple pages concurrently using the Browserless REST API.
// Uses java.net.http.HttpClient (Java 11+) with CompletableFuture — no extra dependencies.
//
// Compile: javac ConcurrentScreenshots.java
// Run:     java ConcurrentScreenshots
//
// Note: for production use, replace the string-concatenated JSON body with a library
// such as Jackson to handle URLs that contain quotes or backslashes safely.

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.stream.IntStream;

public class ConcurrentScreenshots {
    public static void main(String[] args) {
        String token = "YOUR_API_TOKEN_HERE";
        List<String> urls = List.of(
            "https://scraping-sandbox.netlify.app/products",
            "https://scraping-sandbox.netlify.app/contact-us",
            "https://scraping-sandbox.netlify.app/receipt",
            "https://scraping-sandbox.netlify.app/dashboard",
            "https://scraping-sandbox.netlify.app/helix"
        );

        HttpClient client = HttpClient.newHttpClient();

        CompletableFuture<?>[] futures = IntStream.range(0, urls.size())
            .mapToObj(i -> {
                String body = "{\"url\":\"" + urls.get(i) + "\",\"options\":{\"type\":\"png\",\"fullPage\":true}}";
                HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create("https://production-sfo.browserless.io/screenshot?token=" + token))
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(body))
                    .build();

                return client
                    .sendAsync(request, HttpResponse.BodyHandlers.ofByteArray())
                    .thenAccept(res -> {
                        try {
                            Files.write(Path.of("screenshot-" + (i + 1) + ".png"), res.body());
                            System.out.println("Saved screenshot-" + (i + 1) + ".png");
                        } catch (Exception e) {
                            throw new RuntimeException(e);
                        }
                    });
            })
            .toArray(CompletableFuture[]::new);

        CompletableFuture.allOf(futures).join();
        System.out.println("All done");
    }
}
